### PR TITLE
Add roles and permissions to the app

### LINKED ISSUE #21764: New table UI for managing permissions on the policy detail
It's a one to many interface under the hood. Needs to allow you to dynamically configure permissions a la https://app.excalidraw.com/s/DWVAUCmAav/3wkP04RJIoE

### PR DESCRIPTION
**Pull request by hanneskuettner**
## Scope

What's changed:

- The permissions table is now a system interface that can be used for attaching permissions to policies, this enables us to create policies in-line (in a drawer for that matter)
  - The permission details have been moved into that system interface as a separate drawer component
- Added policies to users and roles with the default M2M interface 

## Potential Risks / Drawbacks

- Only a not functioning app, no biggie

### Must haves

- [x] Permission store adaption to the new permissions system
  - [x] Use backend util endpoint for determining the merged permissions of a user
  - [x] Identifying usages of the old permissions
  - [x] Update usages of the old permissions
- [x] User with minimal app permissions can log in
- [x] Bring back public role configuration with policies instead of permissions
- [x] Decide which shares and presets a user with minimal and recommended app permissions can see, as it currently is a `_eq: $CURRENT_ROLE`
- [x] Decide if the `AppUser`, `ActionPermission` and `CollectionPermission` types should be moved to `@directus/types` or be duplicated in `@directus/extensions` for `defineModule`

### Could haves

- [ ] Display role tree in `roles/collection.vue`
- [ ] Show merged permissions in a custom m2m policies interface
- [ ] Show nested users under their parent role?

## (Incomplete) List of breaking changes

- The `app_access`, `admin_access` and `enforce_tfa` flags moved from `User.role` to the top level of a new `AppUser` that is used throughout the app
- The `preRegisterCheck` of a module is called with this new `AppUser` and a record of all user permissions in a condensed form (`Record<string, CollectionPermission>`) instead of a list of all permissions. To identify if a user has access to an action in a collection look for `permissions[collection]?.[action]`, which has the `access`, `full_access` and `fields` properties.

---

 Fixes #21764

### CODE REVIEW

File: app/src/components/v-select/v-select.vue [hunk 1]
```diff
@@ -132,7 +132,7 @@ function useItems() {
 				selectable: get(item, props.itemSelectable),
 				children: children
 					? children.filter((childItem: Record<string, any>) =>
-							filterItem(get(childItem, props.itemText), get(childItem, props.itemValue), childItem.children),
```
Comment (hanneskuettner) [inline app/src/components/v-select/v-select.vue:135]: This is an unrelated fix. `children` at this point is already parsed and transformed to a canonical item form, which now only has the `text` and `value` props, and not the custom ones that one could provide through `itemText` and `itemValue`.

**Review by DanielBiegler — COMMENTED** [review]
_( Just commenting for stuff I find along the way so we can check it off )_

Currently I cant attach users inside the roles view:

```
api dev: [15:22:10] PATCH /roles/43c9f03c-527a-4287-939b-baa46c6a0e70 500 46ms
api dev: [15:22:10.348] ERROR: select `admin_access` from `directus_roles` where `id` = '43c9f03c-527a-4287-939b-baa46c6a0e70' limit 1 - SQLITE_ERROR: no such column: admin_access
api dev:     err: {
api dev:       "type": "Error",
api dev:       "message": "select `admin_access` from `directus_roles` where `id` = '43c9f03c-527a-4287-939b-baa46c6a0e70' limit 1 - SQLITE_ERROR: no such column: admin_access",
api dev:       "stack":
api dev:           Error: select `admin_access` from `directus_roles` where `id` = '43c9f03c-527a-4287-939b-baa46c6a0e70' limit 1 - SQLITE_ERROR: no such column: admin_access
api dev:       "errno": 1,
api dev:       "code": "SQLITE_ERROR",
api dev:       "extensions": {
api dev:         "stack":
api dev:             Error: select `admin_access` from `directus_roles` where `id` = '43c9f03c-527a-4287-939b-baa46c6a0e70' limit 1 - SQLITE_ERROR: no such column: admin_access
api dev:       }
api dev:     }
```

This is because of `checkForOtherAdminUsers` inside of the `updateOne` of roles

```ts
	// TODO move to policies
	private async checkForOtherAdminUsers(
		key: PrimaryKey,
		users: Alterations<User, 'id'> | (string | Partial<User>)[],
	): Promise<void> {
		const role = await this.knex.select('admin_access').from('directus_roles').where('id', '=', key).first();
```

**Comment by hanneskuettner** [comment]
Good catch, actually working on it as we speak :)

File: app/src/interfaces/_system/system-permissions/system-permissions.vue [hunk 2]
```diff
@@ -0,0 +1,714 @@
...
+
+			updateValue();
+		}
+
+		function updateValue() {
+			target.value = cloneDeep(target.value);
+		}
+	}
+
+	function useUtil() {
```
Comment (rijkvanzanten) [inline app/src/interfaces/_system/system-permissions/system-permissions.vue:366]: Wondering if we should extract some of these useX functions to separate ts files to organize this massive setup function a bit

File: app/src/modules/settings/components/navigation.vue [hunk 3]
```diff
@@ -27,9 +27,14 @@ const links = computed<Link[][]>(() => [
 		},
 		{
 			icon: 'admin_panel_settings',
-			name: t('settings_permissions'),
+			name: t('settings_roles'),
 			to: `/settings/roles`,
 		},
+		{
+			icon: 'badge',
```
Comment (rijkvanzanten) [inline app/src/modules/settings/components/navigation.vue:34]: Lets use `admin_panel_settings` as the icon here

```suggestion
			icon: 'admin_panel_settings',
```

File: app/src/modules/settings/components/navigation.vue [hunk 4]
```diff
@@ -27,9 +27,14 @@ const links = computed<Link[][]>(() => [
 		},
 		{
 			icon: 'admin_panel_settings',
```
Comment (rijkvanzanten) [inline app/src/modules/settings/components/navigation.vue:29]: Lets use `group` here

File: app/src/modules/settings/routes/policies/add-new.vue [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+const isOpen = useDialogRoute();
+
+const name = ref<string | null>(null);
+const appAccess = ref(false);
+const adminAccess = ref(false);
+
+const { saving, save } = useSave();
+
+function useSave() {
```
Comment (rijkvanzanten) [inline app/src/modules/settings/routes/policies/add-new.vue:22]: Lets extract this to a `use-save.ts`

File: app/src/modules/settings/routes/policies/collection.vue [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+import { unexpectedError } from '@/utils/unexpected-error';
+import SearchInput from '@/views/private/components/search-input.vue';
+import { Policy } from '@directus/types';
+import { computed, ref } from 'vue';
+import { useI18n } from 'vue-i18n';
+import { useRouter } from 'vue-router';
+import SettingsNavigation from '../../components/navigation.vue';
+
+// Taken from 20240328A-permissions-policies.ts
+const PUBLIC_POLICY_ID = 'abf8a154-5b1c-4a46-ac9c-7300570f4f17';
```
Comment (rijkvanzanten) [inline app/src/modules/settings/routes/policies/collection.vue:14]: We should move / pull this from a single shared constant in `@directus/constants`

**Review by rijkvanzanten — COMMENTED** [review]
Code lgtm

**Comment by hanneskuettner** [comment]
Any input on the open todos? (And still some todos left in code)

**Comment by rijkvanzanten** [comment]
@hanneskuettner Lets tackle shares in a different PR, as there's both app and API work to consider. Lets leave the types as is for now, we can always move them to `types` if we need to reuse them out of the app context

**Review by qugkyp-vudgaz-2maQba — COMMENTED** [review]
Looks great [](https://github.com/directus/directus/pull/22654/files)
